plaso/parsers/czip.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+"""This file contains a parser for compound ZIP files."""
+
+from __future__ import unicode_literals
+
+import struct
+import zipfile
+
+from plaso.lib import errors
+from plaso.parsers import interface
+from plaso.parsers import logger
+from plaso.parsers import manager
+
+
+class CompoundZIPParser(interface.FileObjectParser):
+  """Handles compound ZIP files and provides a slim interface to list and read
+  files from it, acting as a proxy.
+
+  The Plaso engine calls parsers by their Parse() method. The Parse() function
+  checks if the ZIP is valid and calls plugins (CompoundZIPPlugin) registered
+  through the interface by their Process() to produce event objects.
+
+  Plugins are how this parser understands the content inside a compound ZIP
+  file, each plugin holds logic specific to a particular archive file. See the
+  czip_plugins / directory for examples of how czip plugins are implemented.
+  """
+
+  NAME = 'czip'
+  DESCRIPTION = 'Parser for compound ZIP files.'
+
+  _plugin_classes = {}
+
+  def __init__(self):
+    super(CompoundZIPParser, self).__init__()
+    self._file_object = None
+
+  def ParseFileObject(self, parser_mediator, file_object):
+    """Parses a compound ZIP file-like object.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      file_object (dfvfs.FileIO): a file-like object.
+
+    Raises:
+      UnableToParseFile: when the file cannot be parsed.
+    """
+    display_name = parser_mediator.GetDisplayName()
+
+    if not zipfile.is_zipfile(file_object):
+      raise errors.UnableToParseFile(
+          '[{0:s}] unable to parse file: {1:s} with error: {2:s}'.format(
+              self.NAME, display_name, 'Not a Zip file.'))
+
+    # Some non-ZIP files pass the first test but will fail with a negative
+    # seek (IOError) or another error.
+    try:
+      zip_file = zipfile.ZipFile(file_object, 'r')
+      zip_file.close()
+    except (zipfile.BadZipfile, struct.error, zipfile.LargeZipFile):
+      raise errors.UnableToParseFile(
+          '[{0:s}] unable to parse file: {1:s} with error: {2:s}'.format(
+              self.NAME, display_name, 'Bad Zip file.'))
+
+    self._file_object = file_object
+
+    for plugin in self._plugins:
+      try:
+        plugin.UpdateChainAndProcess(parser_mediator, archive_proxy=self)
+      except errors.WrongCompoundZIPPlugin as exception:
+        logger.debug('[{0:s}] wrong plugin: {1!s}'.format(
+            self.NAME, exception))
+
+  def Read(self, file_name):
+    """Reads a given file from the compound ZIP archive.
+
+    Args:
+      file_name (string): file to read
+
+    Returns:
+      string: the content of the file
+    """
+    with zipfile.ZipFile(self._file_object, 'r') as zip_file:
+      return zip_file.read(file_name)
```

Comment:
I'm a bit worried about this being called on a large zipped file here - can you add a check that the file isn't too large before the read is called here?

Reply:
I thought this was enough, it's the way the old parser used to do it https://github.com/SamuelePilleri/plaso/blob/e19fb6c5c10dd63c6c4e015708c127a0e17aba63/plaso/parsers/czip.py#L60

Reply:
I've found a question on StackOverflow that seems to address this: [When are files too large to be read as strings in Python?](https://stackoverflow.com/questions/40894264/when-are-files-too-large-to-be-read-as-strings-in-python) Quoting: _"The theoretical limit is the maximum size of a Python string, which is determined by it's index, which is `2 ** 63` [...] (the theoretical limit is smaller depending on the virtual memory available: that will be around 2GB by default for 32bit python.)"_ and _"if holding a string in memory demands more memory than your system has then you will get a `MemoryError`"_. Also, [Python documentation for `zipfile.read()`](https://docs.python.org/2/library/zipfile.html#zipfile.ZipFile.read) doesn't specify any exception to cover this specific case and I've already checked that the zip file itself isn't too large. All I can do here is blindly catch all exceptions that line could generate and wrap the error message in a more descriptive string.

A possible, but more restrictive, check would be to test if [uncompressed file size](https://docs.python.org/2/library/zipfile.html#zipfile.ZipInfo.file_size) is smaller than 2GB: in fact, the [docs](https://docs.python.org/2/library/zipfile.html#zipfile.ZipFile) specify that ZIP64 extensions are disabled by default: that would be safe on 32 bit systems, but if someone ever enables such extensions, then said check should be dropped.

Let me know what do you prefer.

Reply:
"Too large" is a bit arbitrary, unfortunately. We limit worker memory usage to 4GiB https://github.com/log2timeline/plaso/blob/master/plaso/cli/helpers/process_resources.py#L35 and something like reading a 2GiB file and trying to parse it with a JSON or CSV parser will exceed that pretty easily, and there are plenty of other ways to end up duplicating a 1GB string a few times and exceeding a limit or just being generally slow. 

Catching exceptions wont work by the way, the Plaso foreman will kill the worker process for exceeding the limit.

Good point about Zip64 - please set it on explicitly as it's on by default in Python 3. Also, I assume Facebook dumps can be >2GiB, and that's the main motivation for this change...

Reply:
BTW, https://github.com/log2timeline/dfvfs/blob/master/dfvfs/file_io/zip_file_io.py for the dfVFS implementation of reading a zip file without exploding memory. It's pretty complicated, but might be a way to get a Read() primitive safely.

Reply:
I wonder if I can use it in place of `zipfile`...

Reply:
A perfect implementation would, given a filename which is present inside the zip, return a file-like object from which the JSON (or XML or whatever) parser can read... Do you think this solves our problem? https://docs.python.org/2/library/zipfile.html#zipfile.ZipFile.open

Reply:
Ok, I need to know, is [`plugin.UpdateChainAndProcess()`](https://github.com/SamuelePilleri/plaso/blob/e19fb6c5c10dd63c6c4e015708c127a0e17aba63/plaso/parsers/czip.py#L69) somewhat asynchronous? Because if it's not (or if context manager can handle it) I could just write:

```python
with zipfile.ZipFile(file_object, 'r') as zip_file_object:
  plugin.UpdateChainAndProcess(parser_mediator, zip_file_object)
```

And that is smooooth!

Reply:
Resolving this conversation, current implementation leaves it up to the plugin to decide to either `read()` a whole file inside the archive or `open()` it and handle it more carefully. As for the XML/JSON part, I don't think there will ever be individual files larger than 2 GiB and string copy shouldn't be a problem since the XML/JSON parser can directly read from the archive as the algorithm incrementally deflates it.